zerver/models.py

```diff
@@ -3069,19 +3069,33 @@ def is_status_message(content: str, rendered_content: str) -> bool:
     class Meta:
         indexes = [
             GinIndex("search_tsvector", fastupdate=False, name="zerver_message_search_tsvector"),
-            models.Index(Upper("subject"), name="upper_subject_idx"),
-            models.Index("date_sent", name="zerver_message_date_sent_3b5b05d8"),
             models.Index(
-                "recipient",
+                "realm_id", "recipient_id", "id", name="zerver_message_realm_recipient_id"
+            ),
```

Comment:
So the one other index that we use in the `Message` table is full-text search, which I think isn't declared in the Django files here at all, but instead in `zerver/migrations/0001_initial.py`:

```
        migrations.AddField(                                                                            
            model_name="message",                                                                       
            name="search_tsvector",                                                                     
            field=SearchVectorField(null=True),                                                         
        ),                                                                                              
        migrations.AddIndex(                                                                            
            model_name="message",                                                                       
            index=GinIndex(                                                                             
                "search_tsvector", fastupdate=False, name="zerver_message_search_tsvector"              
            ),                                                                                          
        ),                                                                                              
```

It's an important migration for the long-term to change that index structure to be `(realm, search_tsvector)`; I don't know whether it makes sense to do that while we're doing this or plan to do it separately. 

But one question that thinking about that brought to mind for me is that I think this PR doesn't change the SQLAlchemy code in  `NarrowBuilder` for constructing queries, and that may be nontrivial to adapt? Or maybe it won't be; not sure.

But I think the safer answer might be to avoid dropping the old indexes and instead just put them into a "legacy" section here for now, and only drop them once we've reworked the SQLAlchemy part of the logic to use the new indexes too.

Reply:
Full-text search is more complicated, because it's a GIN index (and it is defined in the Django `zerver/models.py`).  GIN idnexes don't work by default with plain `int` columns inless you add the [`gin_btree`](https://www.postgresql.org/docs/current/btree-gin.html) extension.  And it's not clear how well a multi-column GIN index will actually perform (ref [here](https://pganalyze.com/blog/gin-index), [incident report linked from that](https://gitlab.com/gitlab-com/gl-infra/production/-/issues/4725#note_596146675), [further developer thread](https://www.postgresql.org/message-id/flat/56B332B6.1040109%40promani.be)).

I've pushed a version which adds the index, but we will want to test it carefully to see how it performs in practice.

Reply:
The Ubuntu 20.04 test failed because it takes PostgreSQL 13 or higher to install the `btree_gin` extension without superuser privileges.  This is fixable for testing, but I don't know that there's a simple solution for production upgrades if they're on PostgreSQL 12 or lower.

Reply:
I pulled off the multi-column GIN indexes, since we're having trouble getting them to add concurrently in production, in addition to not having an upgrade path for PostgreSQL 12.

Reply:
OK. Postgres 12 has EOL in about a year, so there would be some cost to dropping it in this release.